proxylib/memcached/text/parser.go

```diff
@@ -0,0 +1,318 @@
+
+	linefeed := bytes.Index(data, []byte("\r\n"))
+	if linefeed < 0 {
+		log.Infof("Did not receive full first line, asking for 1 byte more")
+		return proxylib.MORE, 1
+	}
+
+	// TODO: iterate over data without copying it to new slices
+	// Tokenizing in memcached is done by spaces: https://github.com/memcached/memcached/blob/master/memcached.c#L2978
+	tokens := bytes.Fields(data[:linefeed])
+
+	if !reply {
+		meta := meta.MemcacheMeta{
+			Command:  tokens[0],
+			IsBinary: false,
+		}
+
+		frameLength := 0
+		noreply := false
+		if p.isCommandRetrieval(meta.Command) {
+			// get, gets, gat, gats
+			if bytes.HasPrefix(meta.Command, []byte("get")) {
+				meta.Keys = tokens[1:]
+			} else if bytes.HasPrefix(meta.Command, []byte("gat")) {
+				meta.Keys = tokens[2:]
+			}
+			frameLength = linefeed + 2
+		} else if p.isCommandStorage(meta.Command) {
+			// storage commands
+			meta.Keys = tokens[1:2]
+			nBytes, err := strconv.Atoi(string(tokens[4]))
+			if err != nil {
+				log.Error("Failed to parse storage payload length")
+				return proxylib.ERROR, 0
+			}
+			// 2 additional bytes for terminating linefeed
+			frameLength = linefeed + 2 + nBytes + 2
+
+			if meta.Command[0] == 'c' { //storage command is "cas"
+				noreply = len(tokens) == 7
+			} else {
+				noreply = len(tokens) == 6
+			}
+		} else if p.isCommandDelete(meta.Command) {
+			meta.Keys = tokens[1:2]
+			noreply = len(tokens) == 3
+			frameLength = linefeed + 2
+		} else if p.isCommandIncrDecr(meta.Command) {
+			meta.Keys = tokens[1:2]
+			noreply = len(tokens) == 4
+			frameLength = linefeed + 2
+		} else if bytes.Equal(meta.Command, []byte("touch")) {
+			meta.Keys = tokens[1:2]
+			noreply = len(tokens) == 4
+			frameLength = linefeed + 2
+		} else if bytes.Equal(meta.Command, []byte("slabs")) ||
+			bytes.Equal(meta.Command, []byte("lru")) ||
+			bytes.Equal(meta.Command, []byte("lru_crawler")) ||
+			bytes.Equal(meta.Command, []byte("stats")) ||
+			bytes.Equal(meta.Command, []byte("version")) ||
+			bytes.Equal(meta.Command, []byte("misbehave")) {
+
+			meta.Keys = [][]byte{}
+			frameLength = linefeed + 2
+		} else if bytes.Equal(meta.Command, []byte("flush_all")) ||
+			bytes.Equal(meta.Command, []byte("cache_memlimit")) {
+			meta.Keys = [][]byte{}
+			noreply = bytes.Equal(tokens[len(tokens)-1], []byte("noreply"))
+			frameLength = linefeed + 2
+		} else if bytes.Equal(meta.Command, []byte("quit")) {
+			meta.Keys = [][]byte{}
+			noreply = true
+			frameLength = linefeed + 2
+		} else if bytes.Equal(meta.Command, []byte("watch")) {
+			meta.Keys = [][]byte{}
+			frameLength = linefeed + 2
+			p.watching = true
+		} else {
+			log.Error("Could not parse text memcache frame")
+			return proxylib.ERROR, 0
+		}
+		logEntry := &cilium.LogEntry_GenericL7{
+			&cilium.L7LogEntry{
+				Proto: "textmemcached",
+				Fields: map[string]string{
+					"command": string(meta.Command),
+					"keys":    string(bytes.Join(meta.Keys, []byte(", "))),
+				},
```

Comment:
is the set of access logging key-value pairs significantly different between text and binary variants? If not then could use a common `Proto` for both.

Reply:
I think they are different enough to use different `Proto`.